src/sage/schemes/curves/plane_curve_arrangement.py

```diff
@@ -0,0 +1,1371 @@
+
+    EXAMPLES::
+
+        sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+        sage: H(x, y^2, x-1, y-1)
+        Arrangement (x, y^2, x - 1, y - 1) in Affine Space
+        of dimension 2 over Rational Field
+    """
+    Element = PlaneCurveArrangementElement
+
+    @staticmethod
+    def __classcall__(cls, base, names=()):
+        """
+        Normalize names
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: K = AffinePlaneCurveArrangements(QQ, names=('x', 'y'))
+            sage: H is K
+            True
+            sage: TestSuite(K).run()
+        """
+        names = normalize_names(len(names), names)
+        return super().__classcall__(cls, base, names)
+
+    def __init__(self, base_ring, names=tuple()):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: TestSuite(H).run()
+            sage: H.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: K = ProjectivePlaneCurveArrangements(QQ, names=('x', 'y', 'z'))
+            sage: H is K
+            True
+            sage: type(K)
+            <class 'sage.schemes.curves.plane_curve_arrangement.ProjectivePlaneCurveArrangements_with_category'>
+        """
+        if base_ring not in _Fields:
+            raise ValueError('base ring must be a field')
+        super().__init__(base_ring, names=names, category=Sets())
+        self._embedded = None
+        if len(names) == 2:
+            self._embedded = 'affine'
+        elif len(names) == 3:
+            self._embedded = 'projective'
+
+    def coordinate_ring(self):
+        """
+        Return the coordinate ring.
+
+        OUTPUT:
+
+        The coordinate ring of the curve arrangement.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.coordinate_ring()
+            Multivariate Polynomial Ring in x, y over Rational Field
+        """
+        return PolynomialRing(self.base_ring(), self.variable_names())
+
+    @abstract_method
+    def change_ring(self, base_ring):
+        """
+        Return curve arrangements over a different base ring.
+
+        INPUT:
+
+        - ``base_ring`` -- a ring; the new base ring.
+
+        OUTPUT:
+
+        A new :class:`PlaneCurveArrangements` instance over the new
+        base ring.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = PlaneCurveArrangements(QQ)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: <abstract method ambient_space at  0x...>
+        """
+
+    @abstract_method
+    def ambient_space(self):
+        """
+        Return the ambient space.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = PlaneCurveArrangements(QQ)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: <abstract method ambient_space at  0x...>
```

Comment:
I would write a test using one of the concrete classes.

Reply:
There are already tests in the concrete classes.

Reply:
Yes, but this doesn't even show the method in how it should be used. The error just happens to show it, but that isn't an example either. Having an explicit concrete implementation of the method is useful to a developer showing how it should work.